### PR TITLE
#5 Add Component diagram for Rental service

### LINKED ISSUE #5: Add Component diagram (C3) for Rentals service
**As a** developer,
**I want** to have a Component diagram for Rentals service with some description
**so that** other maintainers and myself have knowledge about service workflow and an agreement about its relations to other services.

### PR DESCRIPTION
**Pull request by RafKulas**
Closes #5

### CODE REVIEW

**Comment by RafKulas** [comment]
image

File: c4builder/src/3 Component Diagrams/3 Rental service/rental-service.md [hunk 1]
```diff
@@ -0,0 +1,3 @@
+**Level 3: Component diagram for *Rental* service**
+
+*Rental* service is responsible for all actions related to the rental of books. The service saves each borrowing as an event into the database and the worker creates a projection that facilitates reading of the database.
```
Comment (krzysztofzajaczkowski) [inline c4builder/src/3 Component Diagrams/3 Rental service/rental-service.md]: Reading `from` the database.

File: c4builder/src/3 Component Diagrams/3 Rental service/rental-service.puml [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+
+Container(gateway, "API Gateway", ".NET with ASP.NET Core", "Routes all requests to appropriate services endpoints", $sprite="dotnet")
+ContainerQueue(rabbit, "Event bus", "RabbitMQ", "Event bus for asynchronous communication", $sprite=rabbitmq)
+ContainerDb(rentals_db, "Rentals database", "PostgreSQL", "Holds rentals events and physical copies information", $sprite="postgresql")
+ContainerDb(rentals_read_db, "Rentals (READ) database", "PostgreSQL", "Holds aggregated rentals and physical copies availability information", $sprite="postgresql")
+
+Container_Boundary(rentals, "Rental service") {
+  Component(rentalController, "Rentals Controller", "Rest Controller", "Allows users to send requests related to book renting.")
+  Component(rentalService, "Rentals Service", "Events publisher", "Produces rentals events.")
+  Component(rentalReader, "Rentals Reader", "Spring service", "Reads and aggregates events from database.")
```
Comment (krzysztofzajaczkowski) [inline c4builder/src/3 Component Diagrams/3 Rental service/rental-service.puml]: `Rentals Reader` doesn't aggregate and/or read events. `Rentals Worker` is responsible for aggregation (projection) and saving projection to READ database. `Rentals Reader` reads projections from READ database.

File: c4builder/src/3 Component Diagrams/3 Rental service/rental-service.puml [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+  Component(rentalWorker, "Rentals Worker", "Events consumer", "Consumes rentals events.")
+  Component(rentalRepository, "Rentals Repository", "Spring repository", "Provides functionality related to database operations like reading and writing events to database")
```
Comment (krzysztofzajaczkowski) [inline c4builder/src/3 Component Diagrams/3 Rental service/rental-service.puml]: `Rentals Repository` shouldn't have any knowledge about events. For READ database it should only know the concept of projected/aggregated information and provide functionality for reading/updating it.

**Review by krzysztofzajaczkowski — CHANGES_REQUESTED** [review]

**Comment by krzysztofzajaczkowski** [comment]
Minor detail, in my opinion commits should be read as a command like sentence, e.g. "Add <xxx> service" (and I think it's generally accepted strategy). Adhering to that rule, commit message should begin with capital letter. If You feel it should be different, we need to discuss it and decide on convention.

**Comment by RafKulas** [comment]
> Minor detail, in my opinion commits should be read as a command like sentence, e.g. "Add service" (and I think it's generally accepted strategy). Adhering to that rule, commit message should begin with capital letter. If You feel it should be different, we need to discuss it and decide on convention.

Yup, we need to discuss convention about commits, branch names and PR names

**Pull request closed without merging**